<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>project.utils.docx &#8212; rsv-neqas 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css?v=27fed22d" />
    <script src="../../../_static/documentation_options.js?v=d45e8c67"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for project.utils.docx</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">docx.py</span>
<span class="sd">=======</span>
<span class="sd">This utilities module is in charge of processing the data of a distribution and preparing its docx report.</span>

<span class="sd">Functions:</span>

<span class="sd">    generate_sample_plot_pdf(sample_name, sample_data, role)</span>
<span class="sd">        Creates a stacked bar and line graph for sample data visualization.</span>

<span class="sd">    create_pygenometracks_plot(reference_genome, annotation, region, bed_path, bigwig_file, bigwig_consensus_file, output_dir, sample_name, user_lab)</span>
<span class="sd">        Uses pyGenomeTracks to generate genome coverage plots.</span>

<span class="sd">    create_element(name)</span>
<span class="sd">        Creates an XML element for DOCX formatting.</span>

<span class="sd">    create_attribute(element, name, value)</span>
<span class="sd">        Adds an attribute to an XML element.</span>

<span class="sd">    add_page_number(run)</span>
<span class="sd">        Inserts a page number field in a DOCX document.</span>
<span class="sd">    </span>
<span class="sd">    generate_docx_reportreport_data, base_dir, role, user_lab, distribution</span>
<span class="sd">        Generates a DOCX report summarizing viric genome analysis results.</span>

<span class="sd">:author: Kevin</span>
<span class="sd">:version: 0.0.1</span>
<span class="sd">:date: 2025-02-21</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">datetime</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="kn">import</span> <span class="n">BytesIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx</span><span class="w"> </span><span class="kn">import</span> <span class="n">Document</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx.shared</span><span class="w"> </span><span class="kn">import</span> <span class="n">Inches</span><span class="p">,</span><span class="n">Pt</span><span class="p">,</span> <span class="n">Cm</span><span class="p">,</span> <span class="n">RGBColor</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx.enum.text</span><span class="w"> </span><span class="kn">import</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="p">,</span> <span class="n">WD_BREAK</span><span class="p">,</span> <span class="n">WD_COLOR_INDEX</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx.enum.table</span><span class="w"> </span><span class="kn">import</span> <span class="n">WD_TABLE_ALIGNMENT</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx.oxml</span><span class="w"> </span><span class="kn">import</span> <span class="n">OxmlElement</span><span class="p">,</span> <span class="n">ns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx.oxml.ns</span><span class="w"> </span><span class="kn">import</span> <span class="n">nsdecls</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">docx.oxml</span><span class="w"> </span><span class="kn">import</span> <span class="n">parse_xml</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">project.utils.report_parser</span><span class="w"> </span><span class="kn">import</span> <span class="n">process_all_reports</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>

<div class="viewcode-block" id="generate_sample_plot_pdf">
<a class="viewcode-back" href="../../../modules.html#project.utils.docx.generate_sample_plot_pdf">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">generate_sample_plot_pdf</span><span class="p">(</span><span class="n">sample_name</span><span class="p">,</span> <span class="n">sample_data</span><span class="p">,</span> <span class="n">role</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a stacked bar and line graph using Matplotlib for a given sample.</span>

<span class="sd">    :param sample_name: The name of the sample, used for file naming.</span>
<span class="sd">    :type sample_name: str</span>
<span class="sd">    :param sample_data: Dictionary containing sample statistics (coverage, similarity, etc.).</span>
<span class="sd">    :type sample_data: dict</span>
<span class="sd">    :param role: User role (affects visualisation style).</span>
<span class="sd">    :type role: str</span>
<span class="sd">    :return: The file path of the generated plot image.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Extract the data for plotting (adjust based on your dataset structure)</span>
    <span class="n">labs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    
    <span class="c1"># Replace &#39;WR006&#39; with &#39;reference&#39; and anonymize the rest of the participants</span>
    <span class="n">anonymized_labs</span> <span class="o">=</span> <span class="p">[</span><span class="n">lab</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">lab</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">labs</span><span class="p">)]</span>  <span class="c1"># [&#39;reference&#39; if lab == &#39;WR024&#39; else f&#39;{i}&#39; for i, lab in enumerate(labs)]</span>
    
    <span class="n">genome_coverage_percent</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span> <span class="k">if</span> <span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">labs</span><span class="p">]</span>  <span class="c1"># Genome coverage at a specific threshold</span>
    <span class="n">similarity_percent</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">labs</span><span class="p">]</span>  <span class="c1"># Could be uniformity or another similarity metric</span>
    <span class="n">num_ns_percent</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span><span class="k">if</span> <span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">labs</span><span class="p">]</span>  <span class="c1"># Total number of non-ACGTNs</span>
    <span class="n">read_coverage</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">sample_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">labs</span><span class="p">]</span>  <span class="c1"># Read depth or other coverage metric</span>

    <span class="c1"># Create the figure and primary axis (ax1) with a wider figure size</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Increased width for a wider figure</span>
    
    <span class="k">if</span> <span class="n">role</span> <span class="o">==</span> <span class="s2">&quot;user&quot;</span><span class="p">:</span>
        <span class="c1"># Horizontal Bar Plot Configuration</span>
        <span class="n">y_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labs</span><span class="p">))</span>
        <span class="n">bar_height</span> <span class="o">=</span> <span class="mf">0.3</span>  <span class="c1"># Limiting bar height (make it a bit thinner)</span>

        <span class="c1"># Plot bars for Genome Coverage and Ns</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">y_pos</span><span class="p">,</span> <span class="n">genome_coverage_percent</span><span class="p">,</span> <span class="n">bar_height</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Genome Coverage (%)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#1E3A5F&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">y_pos</span><span class="p">,</span> <span class="n">num_ns_percent</span><span class="p">,</span> <span class="n">bar_height</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="n">genome_coverage_percent</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Ns in Sequence (%)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#FBC02D&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

        <span class="c1"># Plot a separate bar for Similarity</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">y_pos</span> <span class="o">+</span> <span class="n">bar_height</span><span class="p">,</span> <span class="n">similarity_percent</span><span class="p">,</span> <span class="n">bar_height</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Similarity (%)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#F57C00&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Participant&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Percentage (%)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">y_pos</span> <span class="o">+</span> <span class="n">bar_height</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">anonymized_labs</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">invert_yaxis</span><span class="p">()</span>  <span class="c1"># Invert y-axis for better readability</span>

        <span class="c1"># Secondary axis for Read Coverage</span>
        <span class="n">ax2</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">twiny</span><span class="p">()</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">read_coverage</span><span class="p">,</span> <span class="n">y_pos</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Read Coverage (Mean)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Read Coverage (Mean)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">read_coverage</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Vertical Bar Plot Configuration (Default)</span>
        <span class="n">bar_width</span> <span class="o">=</span> <span class="mf">0.3</span>  <span class="c1"># Limiting bar width (make it a bit narrower)</span>
        <span class="n">x_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labs</span><span class="p">))</span>

        <span class="c1"># Plot bars for Genome Coverage and Ns</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x_pos</span><span class="p">,</span> <span class="n">genome_coverage_percent</span><span class="p">,</span> <span class="n">bar_width</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Genome Coverage (%)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#1E3A5F&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x_pos</span><span class="p">,</span> <span class="n">num_ns_percent</span><span class="p">,</span> <span class="n">bar_width</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="n">genome_coverage_percent</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Ns in Sequence (%)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#FBC02D&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

        <span class="c1"># Plot a separate bar for Similarity</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x_pos</span> <span class="o">+</span> <span class="n">bar_width</span><span class="p">,</span> <span class="n">similarity_percent</span><span class="p">,</span> <span class="n">bar_width</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Similarity (%)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#F57C00&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Participant&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Percentage (%)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">x_pos</span> <span class="o">+</span> <span class="n">bar_width</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">anonymized_labs</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

        <span class="c1"># Secondary axis for Read Coverage</span>
        <span class="n">ax2</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">twinx</span><span class="p">()</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_pos</span><span class="p">,</span> <span class="n">read_coverage</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Read Coverage (Mean)&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Read Coverage (Mean)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">read_coverage</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">)</span>

    <span class="c1"># Add horizontal lines for thresholds</span>
    <span class="k">if</span> <span class="n">role</span> <span class="o">==</span> <span class="s2">&quot;user&quot;</span><span class="p">:</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">90</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;_nolegend_&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">98</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">90</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;_nolegend_&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">98</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Adjust the legend placement to the right of the plot to save vertical space</span>
    <span class="n">handles1</span><span class="p">,</span> <span class="n">labels1</span> <span class="o">=</span> <span class="n">ax1</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
    <span class="n">handles2</span><span class="p">,</span> <span class="n">labels2</span> <span class="o">=</span> <span class="n">ax2</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
    
    <span class="k">if</span> <span class="n">role</span> <span class="o">==</span> <span class="s2">&quot;user&quot;</span><span class="p">:</span>
        <span class="c1"># Combine the legends and place them to the right with a larger font size</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">handles1</span> <span class="o">+</span> <span class="n">handles2</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels1</span> <span class="o">+</span> <span class="n">labels2</span><span class="p">,</span> 
                <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">handles1</span> <span class="o">+</span> <span class="n">handles2</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels1</span> <span class="o">+</span> <span class="n">labels2</span><span class="p">,</span> 
               <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper center&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.27</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>
        
    <span class="c1"># Adjust the layout to prevent clipping of labels and legends</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>  <span class="c1"># Increase padding to avoid overlapping and give extra space to the legend</span>

    <span class="c1"># Save the plot as a PNG image (or you can save as PDF if required)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s1">_metrics_plot.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>  <span class="c1"># Ensure tight bounding box</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s1">_metrics_plot.png&#39;</span>  <span class="c1"># Path to the saved plot image</span></div>



<div class="viewcode-block" id="create_pygenometracks_plot">
<a class="viewcode-back" href="../../../modules.html#project.utils.docx.create_pygenometracks_plot">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_pygenometracks_plot</span><span class="p">(</span><span class="n">reference_genome</span><span class="p">,</span> <span class="n">annotation</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">bed_path</span><span class="p">,</span> <span class="n">bigwig_file</span><span class="p">,</span> <span class="n">bigwig_consensus_file</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">sample_name</span><span class="p">,</span> <span class="n">user_lab</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a genome coverage plot using pyGenomeTracks for a specific user_lab.</span>

<span class="sd">    Creates a tracks.ini file that configures:</span>
<span class="sd">      - A reference track (FASTA),</span>
<span class="sd">      - An annotation track (GFF3),</span>
<span class="sd">      - A bigwig track (BigWig file).</span>

<span class="sd">    :param reference_genome: Path to the reference genome FASTA file.</span>
<span class="sd">    :type reference_genome: str</span>
<span class="sd">    :param annotation: Path to the GFF3 annotation file.</span>
<span class="sd">    :type annotation: str</span>
<span class="sd">    :param region: Genomic region to visualize.</span>
<span class="sd">    :type region: str</span>
<span class="sd">    :param bed_path: Path to the BED file with sequence variants.</span>
<span class="sd">    :type bed_path: str</span>
<span class="sd">    :param bigwig_file: Path to the BigWig file containing read coverage data.</span>
<span class="sd">    :type bigwig_file: str</span>
<span class="sd">    :param bigwig_consensus_file: Path to the consensus BigWig file.</span>
<span class="sd">    :type bigwig_consensus_file: str</span>
<span class="sd">    :param output_dir: Directory where the plot and ini file will be saved.</span>
<span class="sd">    :type output_dir: str</span>
<span class="sd">    :param sample_name: Sample identifier.</span>
<span class="sd">    :type sample_name: str</span>
<span class="sd">    :param user_lab: User lab identifier.</span>
<span class="sd">    :type user_lab: str</span>
<span class="sd">    :return: File path of the generated coverage plot.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
    
    <span class="n">tracks_ini</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s2">_tracks.ini&quot;</span><span class="p">)</span>
    <span class="n">plot_output</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s2">_coverage_plot.png&quot;</span><span class="p">)</span>

    <span class="c1"># Create tracks.ini configuration with three sections:</span>
    <span class="c1"># Reference, Annotation, and BigWig</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tracks_ini</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ini</span><span class="p">:</span>
        <span class="n">ini</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        [annotation]</span>
<span class="s2">        file = </span><span class="si">{</span><span class="n">annotation</span><span class="si">}</span>
<span class="s2">        title = </span><span class="si">{</span><span class="n">reference_genome</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> genes</span>
<span class="s2">        prefered_name = gene_name</span>
<span class="s2">        color = green</span>
<span class="s2">        style = UCSC</span>
<span class="s2">        height = 2</span>
<span class="s2">        file_type = gtf</span>
<span class="s2">        merge_transcripts = true</span>

<span class="s2">        [bed]</span>
<span class="s2">        file = </span><span class="si">{</span><span class="n">bed_path</span><span class="si">}</span>
<span class="s2">        title = Seq. variants</span>
<span class="s2">        color = purple</span>
<span class="s2">        height = 3</span>
<span class="s2">        file_type = bed</span>

<span class="s2">        [bigwig]</span>
<span class="s2">        file = </span><span class="si">{</span><span class="n">bigwig_consensus_file</span><span class="si">}</span>
<span class="s2">        title = Seq. coverage</span>
<span class="s2">        color = grey</span>
<span class="s2">        height = 1</span>
<span class="s2">        file_type = bigwig</span>



<span class="s2">        [bigwig]</span>
<span class="s2">        file = </span><span class="si">{</span><span class="n">bigwig_file</span><span class="si">}</span>
<span class="s2">        title = Read Coverage</span>
<span class="s2">        color = blue</span>
<span class="s2">        height = 3</span>
<span class="s2">        file_type = bigwig</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span>
    
    <span class="c1"># Run pyGenomeTracks to generate the plot for the given region</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;pyGenomeTracks&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--tracks&quot;</span><span class="p">,</span> <span class="n">tracks_ini</span><span class="p">,</span>
        <span class="s2">&quot;--region&quot;</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span>
        <span class="s2">&quot;--outFileName&quot;</span><span class="p">,</span> <span class="n">plot_output</span><span class="p">,</span>
        <span class="s2">&quot;--dpi&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="mi">300</span><span class="p">)</span>
    <span class="p">]</span>
    
    <span class="k">try</span><span class="p">:</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">plot_output</span>  <span class="c1"># Return the path to the generated image</span>
    <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error generating plot: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span></div>



<div class="viewcode-block" id="create_element">
<a class="viewcode-back" href="../../../modules.html#project.utils.docx.create_element">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_element</span><span class="p">(</span><span class="n">name</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates an XML element for use in DOCX formatting.</span>

<span class="sd">    :param name: The name of the XML element.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :return: The created XML element.</span>
<span class="sd">    :rtype: OxmlElement</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">OxmlElement</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="create_attribute">
<a class="viewcode-back" href="../../../modules.html#project.utils.docx.create_attribute">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_attribute</span><span class="p">(</span><span class="n">element</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adds an attribute to an XML element.</span>

<span class="sd">    :param element: The XML element to modify.</span>
<span class="sd">    :type element: OxmlElement</span>
<span class="sd">    :param name: The attribute name.</span>
<span class="sd">    :type name: str</span>
<span class="sd">    :param value: The attribute value.</span>
<span class="sd">    :type value: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">element</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ns</span><span class="o">.</span><span class="n">qn</span><span class="p">(</span><span class="n">name</span><span class="p">),</span> <span class="n">value</span><span class="p">)</span></div>


<div class="viewcode-block" id="add_page_number">
<a class="viewcode-back" href="../../../modules.html#project.utils.docx.add_page_number">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">add_page_number</span><span class="p">(</span><span class="n">run</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Inserts a page number field in a DOCX document.</span>

<span class="sd">    :param run: A DOCX run object where the page number field is inserted.</span>
<span class="sd">    :type run: docx.text.run.Run</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fldChar1</span> <span class="o">=</span> <span class="n">create_element</span><span class="p">(</span><span class="s1">&#39;w:fldChar&#39;</span><span class="p">)</span>
    <span class="n">create_attribute</span><span class="p">(</span><span class="n">fldChar1</span><span class="p">,</span> <span class="s1">&#39;w:fldCharType&#39;</span><span class="p">,</span> <span class="s1">&#39;begin&#39;</span><span class="p">)</span>

    <span class="n">instrText</span> <span class="o">=</span> <span class="n">create_element</span><span class="p">(</span><span class="s1">&#39;w:instrText&#39;</span><span class="p">)</span>
    <span class="n">create_attribute</span><span class="p">(</span><span class="n">instrText</span><span class="p">,</span> <span class="s1">&#39;xml:space&#39;</span><span class="p">,</span> <span class="s1">&#39;preserve&#39;</span><span class="p">)</span>
    <span class="n">instrText</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;PAGE&quot;</span>

    <span class="n">fldChar2</span> <span class="o">=</span> <span class="n">create_element</span><span class="p">(</span><span class="s1">&#39;w:fldChar&#39;</span><span class="p">)</span>
    <span class="n">create_attribute</span><span class="p">(</span><span class="n">fldChar2</span><span class="p">,</span> <span class="s1">&#39;w:fldCharType&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">)</span>

    <span class="n">run</span><span class="o">.</span><span class="n">_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fldChar1</span><span class="p">)</span>
    <span class="n">run</span><span class="o">.</span><span class="n">_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">instrText</span><span class="p">)</span>
    <span class="n">run</span><span class="o">.</span><span class="n">_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fldChar2</span><span class="p">)</span></div>


<div class="viewcode-block" id="generate_docx_report">
<a class="viewcode-back" href="../../../modules.html#project.utils.docx.generate_docx_report">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">generate_docx_report</span><span class="p">(</span><span class="n">report_data</span><span class="p">,</span> <span class="n">base_dir</span><span class="p">,</span> <span class="n">role</span><span class="p">,</span> <span class="n">user_lab</span><span class="p">,</span> <span class="n">distribution</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a DOCX report summarizing genomic analysis results for a given distribution.</span>

<span class="sd">    The report includes sample statistics, visualizations, and formatted text.</span>

<span class="sd">    :param report_data: Processed data containing genomic viral analysis results.</span>
<span class="sd">    :type report_data: dict</span>
<span class="sd">    :param base_dir: Base directory containing report-related files.</span>
<span class="sd">    :type base_dir: str</span>
<span class="sd">    :param role: User role determining report formatting.</span>
<span class="sd">    :type role: str</span>
<span class="sd">    :param user_lab: User lab identifier.</span>
<span class="sd">    :type user_lab: str</span>
<span class="sd">    :param distribution: The distribution to .</span>
<span class="sd">    :type distribution: str</span>
<span class="sd">    :return: the generated DOCX report.</span>
<span class="sd">    :rtype: docx</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">report_data</span> <span class="o">=</span> <span class="n">process_all_reports</span><span class="p">(</span><span class="n">base_dir</span><span class="p">)</span>
    <span class="n">distribution</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">base_dir</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">samples_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_dir</span><span class="si">}</span><span class="s2">/samples.txt&quot;</span>
    <span class="n">sample_reference_map</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">samples_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">sample_id</span><span class="p">,</span> <span class="n">reference</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">sample_reference_map</span><span class="p">[</span><span class="n">sample_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span>

    <span class="kn">from</span><span class="w"> </span><span class="nn">docx.oxml</span><span class="w"> </span><span class="kn">import</span> <span class="n">OxmlElement</span><span class="p">,</span> <span class="n">ns</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">create_element</span><span class="p">(</span><span class="n">name</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">OxmlElement</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">create_attribute</span><span class="p">(</span><span class="n">element</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="n">element</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ns</span><span class="o">.</span><span class="n">qn</span><span class="p">(</span><span class="n">name</span><span class="p">),</span> <span class="n">value</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">add_page_number</span><span class="p">(</span><span class="n">run</span><span class="p">):</span>
        <span class="n">fldChar1</span> <span class="o">=</span> <span class="n">create_element</span><span class="p">(</span><span class="s1">&#39;w:fldChar&#39;</span><span class="p">)</span>
        <span class="n">create_attribute</span><span class="p">(</span><span class="n">fldChar1</span><span class="p">,</span> <span class="s1">&#39;w:fldCharType&#39;</span><span class="p">,</span> <span class="s1">&#39;begin&#39;</span><span class="p">)</span>

        <span class="n">instrText</span> <span class="o">=</span> <span class="n">create_element</span><span class="p">(</span><span class="s1">&#39;w:instrText&#39;</span><span class="p">)</span>
        <span class="n">create_attribute</span><span class="p">(</span><span class="n">instrText</span><span class="p">,</span> <span class="s1">&#39;xml:space&#39;</span><span class="p">,</span> <span class="s1">&#39;preserve&#39;</span><span class="p">)</span>
        <span class="n">instrText</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;PAGE&quot;</span>

        <span class="n">fldChar2</span> <span class="o">=</span> <span class="n">create_element</span><span class="p">(</span><span class="s1">&#39;w:fldChar&#39;</span><span class="p">)</span>
        <span class="n">create_attribute</span><span class="p">(</span><span class="n">fldChar2</span><span class="p">,</span> <span class="s1">&#39;w:fldCharType&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">)</span>

        <span class="n">run</span><span class="o">.</span><span class="n">_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fldChar1</span><span class="p">)</span>
        <span class="n">run</span><span class="o">.</span><span class="n">_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">instrText</span><span class="p">)</span>
        <span class="n">run</span><span class="o">.</span><span class="n">_r</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fldChar2</span><span class="p">)</span>

    <span class="c1"># Create DOCX document</span>
    <span class="n">doc</span> <span class="o">=</span> <span class="n">Document</span><span class="p">()</span>

    <span class="c1"># Style</span>
    <span class="n">font</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">styles</span><span class="p">[</span><span class="s1">&#39;Normal&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">font</span>
    <span class="n">font</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;Arial&#39;</span>
    <span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
    
    <span class="c1">#changing the page margins</span>
    <span class="n">sections</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">sections</span>
    <span class="k">for</span> <span class="n">section</span> <span class="ow">in</span> <span class="n">sections</span><span class="p">:</span>
        <span class="n">section</span><span class="o">.</span><span class="n">top_margin</span> <span class="o">=</span> <span class="n">Cm</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">section</span><span class="o">.</span><span class="n">bottom_margin</span> <span class="o">=</span> <span class="n">Cm</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">section</span><span class="o">.</span><span class="n">left_margin</span> <span class="o">=</span> <span class="n">Cm</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">section</span><span class="o">.</span><span class="n">right_margin</span> <span class="o">=</span> <span class="n">Cm</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Add UK-NEQAS logo to the header (centered)</span>
    <span class="n">header</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">sections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">header</span>
    <span class="n">paragraph</span> <span class="o">=</span> <span class="n">header</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">logo_run</span> <span class="o">=</span> <span class="n">paragraph</span><span class="o">.</span><span class="n">add_run</span><span class="p">()</span>
    <span class="n">logo_run</span><span class="o">.</span><span class="n">add_picture</span><span class="p">(</span><span class="s2">&quot;project/static/uk-neqas-logo.jpg&quot;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">Inches</span><span class="p">(</span><span class="mf">2.5</span><span class="p">))</span>
    <span class="n">date</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
    <span class="n">date</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">RIGHT</span>
    <span class="n">text_run</span> <span class="o">=</span> <span class="n">date</span><span class="o">.</span><span class="n">add_run</span><span class="p">()</span>
    <span class="n">text_run</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%b </span><span class="si">%d</span><span class="s2">, %Y&quot;</span><span class="p">)</span>  <span class="c1"># For center align of text</span>
    <span class="n">text_run</span><span class="o">.</span><span class="n">style</span> <span class="o">=</span> <span class="s2">&quot;Heading 2 Char&quot;</span>

    <span class="c1"># Add a title to the document</span>
    <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="s1">&#39;Quality Metrics Report&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="c1"># First page</span>
    <span class="c1"># Adding the lorem ipsum text to the document</span>
    <span class="n">figure_count</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">table_count</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">intro</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
    <span class="n">intro_run</span> <span class="o">=</span> <span class="n">intro</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="s2">&quot;Introduction&quot;</span><span class="p">)</span>
    <span class="n">intro_run</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;Thank you for participating in the distribution </span><span class="si">{</span><span class="n">distribution</span><span class="si">}</span><span class="s2"> of UK NEQAS Microbiology pilot External Quality Assessment (EQA).</span>

<span class="s2">Samples for this EQA are distributed by UK NEQAS Microbiology and those with detectable virus are either sequenced inhouse or forwarded to the appropriate laboratory for sequencing according to routine practice.</span>

<span class="s2">A survey of sequencing technology is completed as part of the sequencing result upload process. FASTA, BAM and/or FASTQ files are requested to evaluate RSV sequencing quality. The sequence data submitted is also processed by the EQA to generate a lineage using Nextclade.</span>

<span class="s2">Your individual sequencing quality and lineage assignment report (docx format) are available on the XXXXXX website. If you have any problems accessing your reports then please contact XXXXXXXXXXXXXXXXXX (email address). </span>

<span class="s2">The purpose of this EQA is to assess:</span>
<span class="s2"> ➢  The accuracy of RSV sequencing.</span>
<span class="s2"> ➢  Provide a measurement of the quality of viral sequencing.</span>

<span class="s2">The summary of the EQA participation, marking criteria applied, and the scoring is provided as Appendix 1.</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="n">breaks</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span><span class="o">.</span><span class="n">add_run</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
        <span class="n">breaks</span><span class="o">.</span><span class="n">add_break</span><span class="p">()</span>

    <span class="c1"># Add the copyright of CONFIDENTIAL label</span>
    <span class="n">copyright_paragraph</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
    <span class="n">copyright_run</span> <span class="o">=</span> <span class="n">copyright_paragraph</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="s2">&quot;CONFIDENTIAL. Copyright © UKNEQAS for Microbiology&quot;</span><span class="p">)</span>
    <span class="n">copyright_run</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># Bold for emphasis</span>
    <span class="n">copyright_run</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># Italicize for stylistic effect</span>
    <span class="n">copyright_paragraph</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>  <span class="c1"># Center the text</span>
    <span class="n">copyright_run</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Adjust font size for the copyright</span>
    <span class="c1">#copyright_run.add_break(WD_BREAK.PAGE)</span>




    <span class="k">def</span><span class="w"> </span><span class="nf">subtype_assignment</span><span class="p">(</span><span class="n">user_subtype</span><span class="p">,</span> <span class="n">intended_subtype</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">user_subtype</span><span class="o">==</span><span class="s2">&quot;original&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">intended_subtype</span>
        <span class="k">elif</span> <span class="n">user_subtype</span><span class="o">==</span><span class="s2">&quot;alternative&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;RSV-B&quot;</span> <span class="k">if</span> <span class="n">intended_subtype</span><span class="o">==</span><span class="s2">&quot;RSV-A&quot;</span> <span class="k">else</span> <span class="s2">&quot;RSV-A&quot;</span>
        <span class="k">return</span> <span class="n">user_subtype</span>  






    <span class="n">sample_html_reports</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">report_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">sample_name</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">sample_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sample_html_reports</span><span class="p">:</span>
                <span class="n">sample_html_reports</span><span class="p">[</span><span class="n">sample_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">sample_html_reports</span><span class="p">[</span><span class="n">sample_name</span><span class="p">][</span><span class="n">lab</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">compute_evaluation_data</span><span class="p">(</span><span class="n">sample_html_reports</span><span class="p">,</span> <span class="n">sample_reference_map</span><span class="p">,</span> <span class="n">user_lab</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Computes evaluation data from sample reports.&quot;&quot;&quot;</span>
        
        <span class="n">evaluation_data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;RSV_Subtyping&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Clade&quot;</span><span class="p">:[],</span><span class="s2">&quot;Legacy_clade&quot;</span><span class="p">:[],</span> <span class="s2">&quot;Genome Coverage (%)&quot;</span><span class="p">:</span> <span class="p">[],</span><span class="s2">&quot;Ns in Sequence (%)&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Similarity (%)&quot;</span><span class="p">:</span> <span class="p">[],</span><span class="s2">&quot;Read Coverage (mean)&quot;</span><span class="p">:</span> <span class="p">[]}</span>
        <span class="k">for</span> <span class="n">sample_name</span><span class="p">,</span> <span class="n">lab_data</span> <span class="ow">in</span> <span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">reference_metrics</span> <span class="o">=</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;9999&quot;</span><span class="p">,</span> <span class="p">{})</span>  <span class="c1"># Extract reference metrics</span>
            <span class="n">user_metrics</span> <span class="o">=</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">user_lab</span><span class="p">,</span> <span class="p">{})</span>  <span class="c1"># Extract user metrics</span>
            <span class="c1">#subtype</span>
            <span class="n">intended_subtype</span> <span class="o">=</span> <span class="s2">&quot;RSV-B&quot;</span> <span class="k">if</span> <span class="n">sample_reference_map</span><span class="p">[</span><span class="n">sample_name</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;EPI_ISL_1653999&quot;</span> <span class="k">else</span> <span class="s2">&quot;RSV-A&quot;</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">sample_name</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;intended_subtype &quot;</span><span class="p">,</span> <span class="n">intended_subtype</span><span class="p">)</span>  
            <span class="n">reference_subtype</span> <span class="o">=</span> <span class="n">subtype_assignment</span><span class="p">(</span><span class="n">reference_metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">],</span> <span class="n">intended_subtype</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reference_metrics[&#39;subtype&#39;] &quot;</span><span class="p">,</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">])</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reference_subtype &quot;</span><span class="p">,</span> <span class="n">reference_subtype</span><span class="p">)</span>
            <span class="n">user_subtype</span> <span class="o">=</span> <span class="n">subtype_assignment</span><span class="p">(</span><span class="n">user_metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">],</span> <span class="n">intended_subtype</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;user_metrics[&#39;subtype&#39;] &quot;</span><span class="p">,</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">])</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;user_subtype &quot;</span><span class="p">,</span> <span class="n">user_subtype</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;subtype&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;original&quot;</span><span class="p">:</span>
                        <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">subtyping</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="n">user_subtype</span><span class="p">,</span> <span class="n">intended_subtype</span><span class="p">,</span> <span class="n">reference_subtype</span><span class="p">,</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_subtype</span> <span class="o">==</span> <span class="n">intended_subtype</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;RSV_Subtyping&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subtyping</span><span class="p">)</span>    
            
            <span class="c1">#clade</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">intended_clade</span> <span class="o">=</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s2">&quot;clade&quot;</span><span class="p">]</span>
            <span class="n">reference_clade</span> <span class="o">=</span> <span class="n">intended_clade</span>
            <span class="n">user_clade</span> <span class="o">=</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s2">&quot;clade&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;clade&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">intended_clade</span><span class="p">:</span>
                        <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">clade</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="n">user_clade</span><span class="p">,</span> <span class="n">intended_clade</span><span class="p">,</span> <span class="n">reference_clade</span><span class="p">,</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_clade</span> <span class="o">==</span> <span class="n">intended_clade</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;Clade&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">clade</span><span class="p">)</span>  

            <span class="c1">#g_clade</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">intended_clade</span> <span class="o">=</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s2">&quot;G_clade&quot;</span><span class="p">]</span>
            <span class="n">reference_clade</span> <span class="o">=</span> <span class="n">intended_clade</span>
            <span class="n">user_clade</span> <span class="o">=</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s2">&quot;G_clade&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;G_clade&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">intended_clade</span><span class="p">:</span>
                        <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">g_clade</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="n">user_clade</span><span class="p">,</span> <span class="n">intended_clade</span><span class="p">,</span> <span class="n">reference_clade</span><span class="p">,</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_clade</span> <span class="o">==</span> <span class="n">intended_clade</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;Legacy_clade&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g_clade</span><span class="p">)</span>

            <span class="k">def</span><span class="w"> </span><span class="nf">format_IQR_string</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">):</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">aggregated_metrics</span><span class="p">:</span>  <span class="c1"># Handle empty list case</span>
                    <span class="k">return</span> <span class="s2">&quot;N/A&quot;</span>

                <span class="n">mean_value</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">))</span>  <span class="c1"># Compute mean and round</span>
                <span class="n">q1</span><span class="p">,</span> <span class="n">q3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">,</span> <span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">75</span><span class="p">])</span>  <span class="c1"># Compute Q1 and Q3</span>
                <span class="n">q1</span><span class="p">,</span> <span class="n">q3</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">q1</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">q3</span><span class="p">)</span>  <span class="c1"># Round percentiles</span>
                
                <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mean_value</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">q1</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">q3</span><span class="si">}</span><span class="s2">)&quot;</span>

            <span class="c1">#genome_coverage (these four loops could be better refactored into one)</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">reference_coverage</span> <span class="o">=</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">*</span><span class="mi">100</span>
            <span class="n">user_coverage</span> <span class="o">=</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">*</span><span class="mi">100</span>
            <span class="n">aggregated_metrics</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="n">aggregated_metrics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.90</span><span class="p">:</span>
                        <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">coverage</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">user_coverage</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;90</span><span class="si">% o</span><span class="s2">r higher&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">reference_coverage</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_coverage</span> <span class="o">&gt;</span><span class="mf">0.90</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="n">format_IQR_string</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;Genome Coverage (%)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">coverage</span><span class="p">)</span>

            <span class="c1">#Ns</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">reference_n</span> <span class="o">=</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span>
            <span class="n">user_n</span> <span class="o">=</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span>
            <span class="n">aggregated_metrics</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="n">aggregated_metrics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">])</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span><span class="o">&lt;=</span><span class="mi">2</span><span class="p">:</span>
                        <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">ns</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">user_n</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;2</span><span class="si">% o</span><span class="s2">r lower&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">reference_n</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_n</span> <span class="o">&lt;=</span><span class="mi">2</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="n">format_IQR_string</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;Ns in Sequence (%)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span>

            <span class="c1">#similarity</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">reference_similarity</span> <span class="o">=</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span>
            <span class="n">user_similarity</span> <span class="o">=</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span>
            <span class="n">aggregated_metrics</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="n">aggregated_metrics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">])</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">98</span><span class="p">:</span>
                        <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">similarity</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">user_similarity</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;98</span><span class="si">% o</span><span class="s2">r higher&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">reference_similarity</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_similarity</span> <span class="o">&gt;</span><span class="mi">98</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="n">format_IQR_string</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;Similarity (%)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">similarity</span><span class="p">)</span>

            <span class="c1">#Mean coverage depth</span>
            <span class="n">lab_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">lab_pass</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">reference_coverage</span> <span class="o">=</span> <span class="n">reference_metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span>
            <span class="n">user_coverage</span> <span class="o">=</span> <span class="n">user_metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span>
            <span class="n">aggregated_metrics</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">lab_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span><span class="o">==</span><span class="n">user_lab</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s2">&quot;9999&quot;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lab_count</span><span class="o">+=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                        <span class="n">aggregated_metrics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">])</span>
                        <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">50</span><span class="p">:</span>
                            <span class="n">lab_pass</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">coverage</span><span class="o">=</span><span class="p">[</span><span class="n">sample_name</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">user_coverage</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="s2">&quot;50 or higher&quot;</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">reference_coverage</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="s2">&quot;Pass&quot;</span> <span class="k">if</span> <span class="n">user_similarity</span> <span class="o">&gt;</span><span class="mi">50</span> <span class="k">else</span> <span class="s2">&quot;Fail&quot;</span><span class="p">,</span> <span class="n">format_IQR_string</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab_pass</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">lab_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="p">(</span><span class="n">lab_pass</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="p">)</span><span class="w"> </span><span class="o">//</span><span class="w"> </span><span class="n">lab_count</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">]</span>
            <span class="n">evaluation_data</span><span class="p">[</span><span class="s2">&quot;Read Coverage (mean)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">coverage</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="n">evaluation_data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">evaluation_data</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">add_evaluation_tables</span><span class="p">(</span><span class="n">doc</span><span class="p">,</span> <span class="n">evaluation_data</span><span class="p">,</span> <span class="n">run_id</span><span class="o">=</span><span class="s2">&quot;WR024&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds RSV Evaluation and Sequencing Quality tables to the DOCX report with transposed format.&quot;&quot;&quot;</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">apply_color</span><span class="p">(</span><span class="n">run</span><span class="p">,</span> <span class="n">text</span><span class="p">):</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;Apply color coding for pass/fail.&quot;&quot;&quot;</span>
            <span class="n">run</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">text</span>
            <span class="k">if</span> <span class="n">text</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;pass&quot;</span><span class="p">:</span>
                <span class="n">run</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">color</span><span class="o">.</span><span class="n">rgb</span> <span class="o">=</span> <span class="n">RGBColor</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">128</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Green</span>
            <span class="k">elif</span> <span class="n">text</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;fail&quot;</span><span class="p">:</span>
                <span class="n">run</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">color</span><span class="o">.</span><span class="n">rgb</span> <span class="o">=</span> <span class="n">RGBColor</span><span class="p">(</span><span class="mi">255</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Red</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">apply_background_color</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">color_rgb</span><span class="p">):</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;Apply background color to a table cell.&quot;&quot;&quot;</span>
            <span class="c1"># Create the XML shading element and apply it to the cell</span>
            <span class="n">cell</span><span class="o">.</span><span class="n">_element</span><span class="o">.</span><span class="n">get_or_add_tcPr</span><span class="p">()</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">parse_xml</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;&lt;w:shd </span><span class="si">{}</span><span class="s1"> w:fill=&quot;</span><span class="si">{}</span><span class="s1">&quot;/&gt;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nsdecls</span><span class="p">(</span><span class="s1">&#39;w&#39;</span><span class="p">),</span> <span class="n">color_rgb</span><span class="p">)))</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">apply_font_size</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">size_pt</span><span class="p">):</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;Apply a smaller font size to a table cell.&quot;&quot;&quot;</span>
            <span class="k">for</span> <span class="n">paragraph</span> <span class="ow">in</span> <span class="n">cell</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">run</span> <span class="ow">in</span> <span class="n">paragraph</span><span class="o">.</span><span class="n">runs</span><span class="p">:</span>
                    <span class="n">run</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="n">size_pt</span><span class="p">)</span>

        <span class="n">doc</span><span class="o">.</span><span class="n">add_page_break</span><span class="p">()</span>
        <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Evaluation Report: </span><span class="si">{</span><span class="n">run_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Table 1: RSV Subtyping and Clade Assignment</span>
        <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="s2">&quot;Table 1: RSV subtyping and lineage assignment&quot;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">table1</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_table</span><span class="p">(</span><span class="n">rows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>  <span class="c1"># 7 Columns: Indicator, Specimen ID, Your result, Intended, Reference, Score, Participants</span>
        <span class="n">table1</span><span class="o">.</span><span class="n">style</span> <span class="o">=</span> <span class="s2">&quot;Table Grid&quot;</span>
        <span class="n">table1</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_TABLE_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>

        <span class="c1"># Table 1 Header</span>
        <span class="n">hdr_cells</span> <span class="o">=</span> <span class="n">table1</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cells</span>
        <span class="n">headers</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Indicator&quot;</span><span class="p">,</span> <span class="s2">&quot;Specimen ID&quot;</span><span class="p">,</span> <span class="s2">&quot;Your result&quot;</span><span class="p">,</span> <span class="s2">&quot;Intended Result&quot;</span><span class="p">,</span> <span class="s2">&quot;Reference Lab result&quot;</span><span class="p">,</span> <span class="s2">&quot;Your score&quot;</span><span class="p">,</span> <span class="s2">&quot;Participant with intended results&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">text</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">headers</span><span class="p">):</span>
            <span class="n">hdr_cells</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">text</span>
            <span class="n">hdr_cells</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">hdr_cells</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Slightly smaller font size</span>

        <span class="c1"># Apply background color to the header row (light grey)</span>
        <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">hdr_cells</span><span class="p">:</span>
            <span class="n">apply_background_color</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="s2">&quot;D3D3D3&quot;</span><span class="p">)</span>  <span class="c1"># Light grey background</span>

        <span class="c1"># Populate Table 1 with merged Indicator cells and color rows differently</span>
        <span class="n">indicator_colors</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;RSV_Subtyping&quot;</span><span class="p">:</span> <span class="s2">&quot;FFDDC1&quot;</span><span class="p">,</span>  <span class="c1"># Light pink</span>
            <span class="s2">&quot;Clade&quot;</span><span class="p">:</span> <span class="s2">&quot;D1E8E2&quot;</span><span class="p">,</span>  <span class="c1"># Light teal</span>
            <span class="s2">&quot;Legacy_clade&quot;</span><span class="p">:</span> <span class="s2">&quot;E2D1E8&quot;</span><span class="p">,</span>  <span class="c1"># Light purple</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">indicator</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;RSV_Subtyping&quot;</span><span class="p">,</span> <span class="s2">&quot;Clade&quot;</span><span class="p">,</span> <span class="s2">&quot;Legacy_clade&quot;</span><span class="p">]:</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="n">evaluation_data</span><span class="p">[</span><span class="n">indicator</span><span class="p">]</span>
            <span class="n">first_cell</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">row_color</span> <span class="o">=</span> <span class="n">indicator_colors</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">indicator</span><span class="p">,</span> <span class="s2">&quot;FFFFFF&quot;</span><span class="p">)</span>  <span class="c1"># Default to white if not found</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rows</span><span class="p">):</span>
                <span class="n">row_cells</span> <span class="o">=</span> <span class="n">table1</span><span class="o">.</span><span class="n">add_row</span><span class="p">()</span><span class="o">.</span><span class="n">cells</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">first_cell</span> <span class="o">=</span> <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Store first cell for merging</span>
                    <span class="k">if</span> <span class="n">indicator</span><span class="o">==</span><span class="s2">&quot;RSV_Subtyping&quot;</span><span class="p">:</span>
                        <span class="n">first_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;RSV Subtyping&quot;</span>
                    <span class="k">elif</span> <span class="n">indicator</span><span class="o">==</span><span class="s2">&quot;Clade&quot;</span><span class="p">:</span>
                        <span class="n">first_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;Lineage&quot;</span>
                    <span class="k">elif</span> <span class="n">indicator</span><span class="o">==</span><span class="s2">&quot;Legacy_clade&quot;</span><span class="p">:</span>
                        <span class="n">first_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;Legacy lineage&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_tc</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">first_cell</span><span class="o">.</span><span class="n">_tc</span><span class="p">)</span>  <span class="c1"># Merge vertically</span>

                <span class="n">row_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>  
                <span class="n">apply_color</span><span class="p">(</span><span class="n">row_cells</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">add_run</span><span class="p">(),</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>  

                <span class="c1"># Apply background color and smaller font size to each row under the current indicator</span>
                <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">row_cells</span><span class="p">:</span>
                    <span class="n">apply_background_color</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">row_color</span><span class="p">)</span>  <span class="c1"># Apply unique color per indicator</span>
                    <span class="n">apply_font_size</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="mi">9</span><span class="p">)</span>  <span class="c1"># Set font size to 6 pt for the row</span>

        <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>  <span class="c1"># Spacing before Table 2</span>

        <span class="c1"># Table 2: Sequencing Quality</span>
        <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="s2">&quot;Table 2: Sequencing Quality&quot;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">table2</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_table</span><span class="p">(</span><span class="n">rows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>  <span class="c1"># 8 Columns: Indicator, Specimen ID, Your result, Recommended, Reference, Score, Mean (IQR), Participants meeting threshold</span>
        <span class="n">table2</span><span class="o">.</span><span class="n">style</span> <span class="o">=</span> <span class="s2">&quot;Table Grid&quot;</span>
        <span class="n">table2</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_TABLE_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>

        <span class="c1"># Create the first header row</span>
        <span class="n">table2_header</span> <span class="o">=</span> <span class="n">table2</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cells</span>
        <span class="n">headers</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Indicator&quot;</span><span class="p">,</span> <span class="s2">&quot;Specimen ID&quot;</span><span class="p">,</span> <span class="s2">&quot;Your result&quot;</span><span class="p">,</span> <span class="s2">&quot;Recommended Value*&quot;</span><span class="p">,</span> <span class="s2">&quot;Reference Lab result&quot;</span><span class="p">,</span> <span class="s2">&quot;Your score&quot;</span><span class="p">,</span> <span class="s2">&quot;Participant summary&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">text</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">headers</span><span class="p">):</span>
            <span class="n">table2_header</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">text</span>
            <span class="n">table2_header</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">table2_header</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Slightly smaller font size</span>
        <span class="n">table2_header</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">table2_header</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span>

        <span class="c1"># Apply background color to the header row (light grey)</span>
        <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">table2_header</span><span class="p">:</span>
            <span class="n">apply_background_color</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="s2">&quot;D3D3D3&quot;</span><span class="p">)</span>  <span class="c1"># Light grey background</span>

        <span class="c1"># Add the second empty row</span>
        <span class="n">empty_row</span> <span class="o">=</span> <span class="n">table2</span><span class="o">.</span><span class="n">add_row</span><span class="p">()</span><span class="o">.</span><span class="n">cells</span>

        <span class="c1"># Merge all cells except the one under &quot;Participant Summary&quot; (column 6)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">):</span>  <span class="c1"># Merge the first 6 columns</span>
            <span class="n">empty_row</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">table2_header</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

        <span class="c1"># Split the &quot;Participant Summary&quot; cell into two: &quot;Mean (IQR)&quot; and &quot;Participants Meeting Threshold&quot;</span>
        <span class="n">empty_row</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;Mean (IQR)&quot;</span>
        <span class="n">empty_row</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;Participants meeting threshold&quot;</span>

        <span class="c1"># Optional: Bold the new header cells and make their font smaller</span>
        <span class="n">empty_row</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Smaller font size</span>
        <span class="n">empty_row</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Smaller font size</span>

        <span class="c1"># Apply background color to the empty row (light blue)</span>
        <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">empty_row</span><span class="p">:</span>
            <span class="n">apply_background_color</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="s2">&quot;ADD8E6&quot;</span><span class="p">)</span>  <span class="c1"># Light blue background</span>

        <span class="c1"># Populate Table 2 with merged Indicator cells and color rows differently</span>
        <span class="n">indicator_colors_table2</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Genome Coverage (%)&quot;</span><span class="p">:</span> <span class="s2">&quot;FFDDC1&quot;</span><span class="p">,</span>  <span class="c1"># Light pink</span>
            <span class="s2">&quot;Ns in Sequence (%)&quot;</span><span class="p">:</span> <span class="s2">&quot;D1E8E2&quot;</span><span class="p">,</span>  <span class="c1"># Light teal</span>
            <span class="s2">&quot;Similarity (%)&quot;</span><span class="p">:</span> <span class="s2">&quot;E2D1E8&quot;</span><span class="p">,</span>  <span class="c1"># Light purple</span>
            <span class="s2">&quot;Read Coverage (mean)&quot;</span><span class="p">:</span> <span class="s2">&quot;FFFACD&quot;</span><span class="p">,</span>  <span class="c1"># Light yellow</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">indicator</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;Genome Coverage (%)&quot;</span><span class="p">,</span> <span class="s2">&quot;Ns in Sequence (%)&quot;</span><span class="p">,</span> <span class="s2">&quot;Similarity (%)&quot;</span><span class="p">,</span> <span class="s2">&quot;Read Coverage (mean)&quot;</span><span class="p">]:</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="n">evaluation_data</span><span class="p">[</span><span class="n">indicator</span><span class="p">]</span>
            <span class="n">first_cell</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">row_color</span> <span class="o">=</span> <span class="n">indicator_colors_table2</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">indicator</span><span class="p">,</span> <span class="s2">&quot;FFFFFF&quot;</span><span class="p">)</span>  <span class="c1"># Default to white if not found</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rows</span><span class="p">):</span>
                <span class="n">row_cells</span> <span class="o">=</span> <span class="n">table2</span><span class="o">.</span><span class="n">add_row</span><span class="p">()</span><span class="o">.</span><span class="n">cells</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">first_cell</span> <span class="o">=</span> <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Store first cell for merging</span>
                    <span class="n">first_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">indicator</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">_tc</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">first_cell</span><span class="o">.</span><span class="n">_tc</span><span class="p">)</span>  <span class="c1"># Merge vertically</span>

                <span class="n">row_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>  
                <span class="n">apply_color</span><span class="p">(</span><span class="n">row_cells</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">add_run</span><span class="p">(),</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>  
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>  

                <span class="c1"># Apply background color and smaller font size to each row under the current indicator</span>
                <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">row_cells</span><span class="p">:</span>
                    <span class="n">apply_background_color</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="n">row_color</span><span class="p">)</span>  <span class="c1"># Apply unique color per indicator</span>
                    <span class="n">apply_font_size</span><span class="p">(</span><span class="n">cell</span><span class="p">,</span> <span class="mi">9</span><span class="p">)</span>  <span class="c1"># Set font size to 6 pt for the row</span>
        
        <span class="c1"># **Adding the Sequencing Quality Notes with Smaller Font**</span>
        <span class="n">para</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">* Recommended Value&quot;</span><span class="p">)</span>

        <span class="n">bullet_points</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;Obtained sufficient genome coverage (90</span><span class="si">% o</span><span class="s2">r higher).&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Maintained Ns in Sequence within acceptable limits (2</span><span class="si">% o</span><span class="s2">r lower).&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Obtained sufficient Similarity (98</span><span class="si">% o</span><span class="s2">r higher).&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Obtained sufficient Read Coverage (Mean depth of 50 or higher).&quot;</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">point</span> <span class="ow">in</span> <span class="n">bullet_points</span><span class="p">:</span>
            <span class="n">para</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">point</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s2">&quot;List Bullet&quot;</span><span class="p">)</span>
            <span class="n">para</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">Pt</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>  <span class="c1"># Set font size to 6 pt</span>
        
        <span class="n">doc</span><span class="o">.</span><span class="n">add_page_break</span><span class="p">()</span>


        <span class="k">return</span> <span class="n">doc</span>


    <span class="n">evaluation_data</span> <span class="o">=</span> <span class="n">compute_evaluation_data</span><span class="p">(</span><span class="n">sample_html_reports</span><span class="p">,</span> <span class="n">sample_reference_map</span><span class="p">,</span> <span class="n">user_lab</span><span class="p">)</span>
    <span class="n">doc</span> <span class="o">=</span> <span class="n">add_evaluation_tables</span><span class="p">(</span><span class="n">doc</span><span class="p">,</span> <span class="n">evaluation_data</span><span class="p">,</span> <span class="n">user_lab</span><span class="p">)</span>

    <span class="c1"># If the user is not a superuser, filter and aggregate data</span>
    <span class="n">others_label</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
    <span class="n">agg_users</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">role</span> <span class="o">==</span> <span class="s2">&quot;superuser&quot;</span><span class="p">:</span>
        <span class="n">processed_reports</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">sample_name</span><span class="p">,</span> <span class="n">labs_data</span> <span class="ow">in</span> <span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">user_lab</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">labs_data</span><span class="p">:</span>
                <span class="k">continue</span>  <span class="c1"># Skip samples the user has no access to</span>

            <span class="n">user_metrics</span> <span class="o">=</span> <span class="n">labs_data</span><span class="p">[</span><span class="n">user_lab</span><span class="p">]</span>
            <span class="n">reference_metrics</span> <span class="o">=</span> <span class="n">labs_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;9999&quot;</span><span class="p">,</span> <span class="p">{})</span>  <span class="c1"># Extract reference metrics</span>
            <span class="n">subtype_counts</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">clade_counts</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">g_clade_counts</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="n">aggregated_metrics</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;coverage&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;Ns&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;similarity&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;lab_count&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;clade&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                <span class="s2">&quot;G_clade&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                <span class="s2">&quot;subtype&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;</span>
            <span class="p">}</span>

            <span class="c1"># Aggregate data from other labs</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">labs_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span> <span class="o">!=</span> <span class="n">user_lab</span> <span class="ow">and</span> <span class="n">lab</span> <span class="o">!=</span> <span class="s2">&quot;9999&quot;</span> <span class="ow">and</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>  <span class="c1"># Exclude user lab and reference lab</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                        <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="c1"># Count clade occurrences</span>
                    <span class="n">clade</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;clade&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                    <span class="n">subtype</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;subtype&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                    <span class="n">g_clade</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;G_clade&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">subtype</span><span class="p">:</span>
                        <span class="n">subtype_counts</span><span class="p">[</span><span class="n">subtype</span><span class="p">]</span> <span class="o">=</span> <span class="n">subtype_counts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">subtype</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">clade</span><span class="p">:</span>
                        <span class="n">clade_counts</span><span class="p">[</span><span class="n">clade</span><span class="p">]</span> <span class="o">=</span> <span class="n">clade_counts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">clade</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">g_clade</span><span class="p">:</span>
                        <span class="n">g_clade_counts</span><span class="p">[</span><span class="n">g_clade</span><span class="p">]</span> <span class="o">=</span> <span class="n">g_clade_counts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">g_clade</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>

            <span class="c1"># Calculate averages</span>
            <span class="k">if</span> <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">]),</span> <span class="mi">2</span>
                <span class="p">)</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;Ns&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">],</span> <span class="mi">2</span>
                <span class="p">)</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;similarity&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">],</span> <span class="mi">2</span>
                <span class="p">)</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span>
                    <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">],</span> <span class="mi">2</span>
                <span class="p">)</span>

            <span class="c1"># Determine the most frequent clades</span>
            <span class="k">if</span> <span class="n">subtype_counts</span><span class="p">:</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;subtype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">subtype_counts</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">subtype_counts</span><span class="o">.</span><span class="n">get</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">clade_counts</span><span class="p">:</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;clade&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">clade_counts</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">clade_counts</span><span class="o">.</span><span class="n">get</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">g_clade_counts</span><span class="p">:</span>
                <span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;G_clade&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">g_clade_counts</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">g_clade_counts</span><span class="o">.</span><span class="n">get</span><span class="p">)</span>

            <span class="c1"># Prepare processed data for the user</span>
            <span class="n">others_label</span><span class="o">=</span><span class="s2">&quot;Others (&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">aggregated_metrics</span><span class="p">[</span><span class="s2">&quot;lab_count&quot;</span><span class="p">])</span><span class="o">+</span><span class="s2">&quot;)&quot;</span>
            <span class="n">processed_reports</span><span class="p">[</span><span class="n">sample_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">user_lab</span><span class="p">:</span> <span class="n">user_metrics</span><span class="p">,</span>
                <span class="n">others_label</span><span class="p">:</span> <span class="n">aggregated_metrics</span><span class="p">,</span>
                <span class="s2">&quot;Reference&quot;</span><span class="p">:</span> <span class="n">reference_metrics</span>
            <span class="p">}</span>
            <span class="n">agg_users</span><span class="p">[</span><span class="n">sample_name</span><span class="p">]</span><span class="o">=</span><span class="n">aggregated_metrics</span>

        <span class="n">sample_html_reports</span> <span class="o">=</span> <span class="n">processed_reports</span>


    <span class="n">sample_html_reports</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
    
    <span class="c1"># Handle lab anonymization based on the user&#39;s role</span>
    <span class="k">if</span> <span class="n">role</span> <span class="o">==</span> <span class="s1">&#39;superuser&#39;</span><span class="p">:</span>
        <span class="c1"># Iterate through each sample&#39;s data and anonymize lab names</span>
        <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">anonymized_data</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="k">if</span> <span class="n">lab</span> <span class="o">==</span> <span class="s1">&#39;9999&#39;</span> <span class="ow">or</span> <span class="n">lab</span> <span class="o">==</span> <span class="s1">&#39;Reference&#39;</span><span class="p">:</span>
                    <span class="n">anonymized_data</span><span class="p">[</span><span class="s1">&#39;9999&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics</span>  <span class="c1"># Keep the reference lab as &#39;9999&#39;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">anonymized_data</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s1">01d</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics</span>  <span class="c1"># Anonymize labs as single-digit numbers</span>
                    <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1"># Replace the sample data with anonymized data</span>
            <span class="n">sample_html_reports</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">anonymized_data</span>

    <span class="c1"># Sort sample_html_reports such that &#39;9999&#39; comes first</span>
    <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">sorted_data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">key</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="n">k</span> <span class="o">!=</span> <span class="s1">&#39;9999&#39;</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>  <span class="c1"># &#39;9999&#39; comes first, then rest in order</span>
        <span class="p">}</span>
        <span class="n">sample_html_reports</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">sorted_data</span>

    <span class="c1"># Sort the sample_html_reports dictionary by sample name</span>
    <span class="n">sample_html_reports</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>



    <span class="c1"># Add sample plots and tables to the DOCX file</span>
    <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">sample_html_reports</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plot_path</span> <span class="o">=</span> <span class="n">generate_sample_plot_pdf</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">role</span><span class="p">)</span>  <span class="c1"># Generate and get the plot path</span>
        <span class="n">intended_subtype</span> <span class="o">=</span> <span class="s2">&quot;RSV-B&quot;</span> <span class="k">if</span> <span class="n">sample_reference_map</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;EPI_ISL_1653999&quot;</span> <span class="k">else</span> <span class="s2">&quot;RSV-A&quot;</span>
        <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Sample </span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Lineage assignment&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

        <span class="c1"># Extract Clade and G_clade assignments for the &#39;reference&#39;</span>
        <span class="n">reference_clade</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">reference_g_clade</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Iterate through data to find the reference clade and g_clade</span>
        <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">lab</span> <span class="o">==</span> <span class="s1">&#39;9999&#39;</span><span class="p">:</span>  <span class="c1"># Reference participant</span>
                <span class="n">reference_clade</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;clade&#39;</span><span class="p">]</span>
                <span class="n">reference_g_clade</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;G_clade&#39;</span><span class="p">]</span>
                <span class="k">break</span>
            <span class="k">elif</span> <span class="n">lab</span><span class="o">==</span> <span class="s1">&#39;Reference&#39;</span><span class="p">:</span>
                <span class="n">reference_subtype</span> <span class="o">=</span> <span class="n">subtype_assignment</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">],</span> <span class="n">intended_subtype</span><span class="p">)</span>
                <span class="n">reference_clade</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;clade&#39;</span><span class="p">]</span>
                <span class="n">reference_g_clade</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;G_clade&#39;</span><span class="p">]</span>
                <span class="k">break</span>

        <span class="c1"># Add an introductory paragraph about clade assignments</span>
        <span class="k">if</span> <span class="n">role</span><span class="o">==</span><span class="s1">&#39;superuser&#39;</span><span class="p">:</span>
            <span class="c1">#doc.add_paragraph(&quot;All participants matched RSV subtype assignments with reference lab.&quot;, style=&#39;List Bullet&#39;)</span>
            <span class="n">mistakes_text</span> <span class="o">=</span> <span class="s2">&quot;All participants matched lineage assignments with reference lab.&quot;</span>  <span class="c1"># Default</span>
            <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">lab</span> <span class="o">!=</span> <span class="s1">&#39;9999&#39;</span> <span class="ow">and</span> <span class="n">lab</span><span class="o">!=</span><span class="s1">&#39;Reference&#39;</span><span class="p">:</span>  <span class="c1"># Exclude the reference itself</span>
                    <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;clade&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">reference_clade</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;G_clade&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">reference_g_clade</span><span class="p">:</span>
                        <span class="n">mistakes_text</span> <span class="o">=</span> <span class="s2">&quot;Some participants&#39; clade assignments differed from the reference lab.&quot;</span>
                        <span class="n">mistakes_found</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">break</span>
            <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="n">mistakes_text</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mistakes_text</span> <span class="o">=</span> <span class="s2">&quot;Your lab&#39;s lineage assignment matches the reference lab&#39;s.&quot;</span>
            <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="n">user_lab</span><span class="p">][</span><span class="s1">&#39;clade&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">reference_clade</span><span class="p">:</span>
                <span class="n">mistakes_text</span> <span class="o">=</span> <span class="s2">&quot;Your lab&#39;s lineage assignment does not match the reference lab&#39;s.&quot;</span>
            <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="n">mistakes_text</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
        
        <span class="c1"># Caption for the Clade table</span>
        <span class="n">clade_caption</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
        <span class="n">clade_caption</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
        <span class="n">clade_runner</span> <span class="o">=</span> <span class="n">clade_caption</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Table </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">table_count</span><span class="p">)</span><span class="si">}</span><span class="s2">. Lineage assignments for sample </span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">, including RSV subtype.&quot;</span><span class="p">)</span>
        <span class="n">clade_runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">clade_runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">table_count</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># Increment the table count</span>

        <span class="c1"># Add the Clade and G_clade table with RSV subtype</span>
        <span class="n">clade_table</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_table</span><span class="p">(</span><span class="n">rows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>  <span class="c1"># Creating a table with 4 columns (Participant, Subtype, Clade, Legacy clade)</span>
        <span class="n">clade_table</span><span class="o">.</span><span class="n">style</span> <span class="o">=</span> <span class="s1">&#39;Table Grid&#39;</span>

        <span class="c1"># Adding the header row for the table</span>
        <span class="n">clade_hdr_cells</span> <span class="o">=</span> <span class="n">clade_table</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cells</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Participant&#39;</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;RSV Subtype&#39;</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Lineage&#39;</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Legacy lineage&#39;</span>
        <span class="n">clade_hdr_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Adding data rows for each participant and highlighting mismatches</span>
        <span class="n">labCount</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">labName</span><span class="o">=</span><span class="n">lab</span>

            <span class="c1"># Add a new row to the table</span>
            <span class="n">row_cells</span> <span class="o">=</span> <span class="n">clade_table</span><span class="o">.</span><span class="n">add_row</span><span class="p">()</span><span class="o">.</span><span class="n">cells</span>

            <span class="c1"># Participant column</span>
            <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">labName</span>
            <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># Italicize the first column (Participant)</span>

            <span class="c1"># RSV Subtype column</span>
            <span class="n">subtype_cell</span> <span class="o">=</span> <span class="n">row_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">subtype_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">subtype_assignment</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">],</span><span class="n">intended_subtype</span><span class="p">)</span>
            <span class="c1"># Highlight mismatched Clade values</span>
            <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;subtype&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;alternative&quot;</span><span class="p">:</span>
                <span class="n">subtype_cell</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">highlight_color</span> <span class="o">=</span> <span class="n">WD_COLOR_INDEX</span><span class="o">.</span><span class="n">YELLOW</span>

            <span class="c1"># Clade column</span>
            <span class="n">clade_cell</span> <span class="o">=</span> <span class="n">row_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">clade_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;clade&#39;</span><span class="p">]</span>
            <span class="c1"># Highlight mismatched Clade values</span>
            <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;clade&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">reference_clade</span><span class="p">:</span>
                <span class="n">clade_cell</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">highlight_color</span> <span class="o">=</span> <span class="n">WD_COLOR_INDEX</span><span class="o">.</span><span class="n">YELLOW</span>

            <span class="c1"># Legacy clade column</span>
            <span class="n">g_clade_cell</span> <span class="o">=</span> <span class="n">row_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">g_clade_cell</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;G_clade&#39;</span><span class="p">]</span>
            <span class="c1"># Highlight mismatched Legacy clade values</span>
            <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;G_clade&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">reference_g_clade</span><span class="p">:</span>
                <span class="n">g_clade_cell</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">highlight_color</span> <span class="o">=</span> <span class="n">WD_COLOR_INDEX</span><span class="o">.</span><span class="n">YELLOW</span>


        <span class="n">doc</span><span class="o">.</span><span class="n">add_heading</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Sequencing quality&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

                <span class="c1"># Initialize flags to check if any participants fail to meet thresholds</span>
        <span class="n">failed_coverage</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">failed_ns</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">failed_similarity</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">failed_coverage_depth</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Initialize user-specific flags if the role is not a superuser</span>
        <span class="n">user_failed_coverage</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">user_failed_ns</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">user_failed_similarity</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">user_failed_coverage_depth</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># Loop through the data to check conditions and generate summary statements</span>
        <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">labName</span> <span class="o">=</span> <span class="n">lab</span>

            <span class="c1"># Check thresholds for all participants if the role is superuser</span>
            <span class="k">if</span> <span class="n">role</span> <span class="o">==</span> <span class="s2">&quot;superuser&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.90</span><span class="p">:</span>
                    <span class="n">failed_coverage</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">labName</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="n">failed_ns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">labName</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;similarity&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">98</span><span class="p">:</span>
                    <span class="n">failed_similarity</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">labName</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Mean coverage depth&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="nb">float</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Mean coverage depth&#39;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">50</span><span class="p">:</span>
                    <span class="n">failed_coverage_depth</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">labName</span><span class="p">)</span>

            <span class="c1"># Check thresholds for user lab only if the role is not a superuser</span>
            <span class="k">elif</span> <span class="n">lab</span> <span class="o">==</span> <span class="n">user_lab</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.90</span><span class="p">:</span>
                    <span class="n">user_failed_coverage</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="n">user_failed_ns</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;similarity&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">98</span><span class="p">:</span>
                    <span class="n">user_failed_similarity</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Mean coverage depth&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N/A&#39;</span> <span class="ow">or</span> <span class="nb">float</span><span class="p">(</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Mean coverage depth&#39;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">50</span><span class="p">:</span>
                    <span class="n">user_failed_coverage_depth</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Generate feedback based on role</span>
        <span class="k">if</span> <span class="n">role</span> <span class="o">==</span> <span class="s2">&quot;superuser&quot;</span><span class="p">:</span>
            <span class="c1"># Add summary lines for superuser</span>
            <span class="k">if</span> <span class="n">failed_coverage</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Participants </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">failed_coverage</span><span class="p">)</span><span class="si">}</span><span class="s2"> failed to satisfy the threshold for genome coverage (90%).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;All participants obtained sufficient genome coverage (90</span><span class="si">% o</span><span class="s2">r higher).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">failed_ns</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Participants </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">failed_ns</span><span class="p">)</span><span class="si">}</span><span class="s2"> failed to satisfy the threshold for Ns in Sequence (greater than 2%).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;All participants maintained Ns in Sequence within acceptable limits (2</span><span class="si">% o</span><span class="s2">r lower).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">failed_similarity</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Participants </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">failed_similarity</span><span class="p">)</span><span class="si">}</span><span class="s2"> failed to satisfy the threshold for Similarity (98%).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;All participants obtained sufficient Similarity (98</span><span class="si">% o</span><span class="s2">r higher).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">failed_coverage_depth</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Participants </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">failed_coverage_depth</span><span class="p">)</span><span class="si">}</span><span class="s2"> failed to satisfy the threshold for Read Coverage (Mean depth of 50).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;All participants obtained sufficient Read Coverage (Mean depth of 50 or higher).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Add summary lines for non-superuser</span>
            <span class="k">if</span> <span class="n">user_failed_coverage</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Failed to satisfy the threshold for genome coverage (90%).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Obtained sufficient genome coverage (90</span><span class="si">% o</span><span class="s2">r higher).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">user_failed_ns</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Failed to satisfy the threshold for Ns in Sequence (greater than 2%).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Obtained sufficient Ns in Sequence (2</span><span class="si">% o</span><span class="s2">r lower).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">user_failed_similarity</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Failed to satisfy the threshold for Similarity (98%).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Obtained sufficient Similarity (98</span><span class="si">% o</span><span class="s2">r higher).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">user_failed_coverage_depth</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Failed to satisfy the threshold for Read Coverage (Mean depth of 50).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">(</span><span class="s2">&quot;Obtained sufficient Read Coverage (Mean depth of 50 or higher).&quot;</span><span class="p">,</span> <span class="n">style</span><span class="o">=</span><span class="s1">&#39;List Bullet&#39;</span><span class="p">)</span>
        
        <span class="c1"># Add the plot image to DOCX with a caption</span>
        <span class="n">doc</span><span class="o">.</span><span class="n">add_picture</span><span class="p">(</span><span class="n">plot_path</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">Inches</span><span class="p">(</span><span class="mi">6</span><span class="p">))</span>  <span class="c1"># Adjust size as needed</span>
        <span class="n">last_paragraph</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> 
        <span class="n">last_paragraph</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
        <span class="n">para</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
        <span class="n">runner</span> <span class="o">=</span> <span class="n">para</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Figure </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">figure_count</span><span class="p">)</span><span class="si">}</span><span class="s2">. Quality metrics for sample </span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2"> for participants that submitted appropriate data files&quot;</span><span class="p">)</span>
        <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">last_paragraph</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> 
        <span class="n">last_paragraph</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
        <span class="n">figure_count</span><span class="o">+=</span><span class="mi">1</span>
        
        <span class="c1"># Caption for the table</span>
        <span class="n">para</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
        <span class="n">para</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
        <span class="n">runner</span> <span class="o">=</span> <span class="n">para</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Table </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">table_count</span><span class="p">)</span><span class="si">}</span><span class="s2">. Quality metrics data for sample </span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Generate and insert the metrics table for the sample</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_table</span><span class="p">(</span><span class="n">rows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cols</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># Creating a table with 5 columns for metrics</span>
        <span class="n">table</span><span class="o">.</span><span class="n">style</span> <span class="o">=</span> <span class="s1">&#39;Table Grid&#39;</span>

        <span class="c1"># Adding the header row for the table</span>
        <span class="n">hdr_cells</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cells</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Participant&#39;</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Genome Coverage (%)&#39;</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Ns in Sequence (%)&#39;</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Similarity (%)&#39;</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="s1">&#39;Read Coverage (Mean)&#39;</span>
        <span class="n">hdr_cells</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Add data rows for each lab/sample entry</span>
        <span class="n">labCount</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">lab</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1">#if lab == &#39;WR024&#39;:</span>
            <span class="c1">#    labName = &#39;reference&#39;</span>
            <span class="c1">#else:</span>
            <span class="c1">#labCount += 1</span>
            <span class="c1">#labName = str(labCount)</span>
            <span class="n">labName</span><span class="o">=</span><span class="n">lab</span>

            <span class="c1"># Add a new row to the table</span>
            <span class="n">row_cells</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">add_row</span><span class="p">()</span><span class="o">.</span><span class="n">cells</span>

            <span class="c1"># Fill in the cells for each column</span>
            <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="n">labName</span>  <span class="c1"># First column with lab name (bold)</span>
            <span class="n">row_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># Bold the first column</span>

            <span class="c1"># Coverage column (highlight if less than 90%)</span>
            <span class="n">coverage_value</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span> <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span> <span class="k">else</span> <span class="s2">&quot;N/A&quot;</span>
            <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">coverage_value</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">coverage_value</span> <span class="o">&lt;</span> <span class="mi">90</span><span class="p">:</span>
                    <span class="n">row_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">highlight_color</span> <span class="o">=</span> <span class="n">WD_COLOR_INDEX</span><span class="o">.</span><span class="n">YELLOW</span>  <span class="c1"># Highlight if less than 90%</span>
                <span class="c1"># Ns column (highlight if higher than 2%)</span>
                <span class="n">ns_value</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Ns&#39;</span><span class="p">]</span>
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ns_value</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">ns_value</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="n">row_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">highlight_color</span> <span class="o">=</span> <span class="n">WD_COLOR_INDEX</span><span class="o">.</span><span class="n">YELLOW</span>  <span class="c1"># Highlight if higher than 2%</span>

                <span class="c1"># Similarity column (highlight if less than 98%)</span>
                <span class="n">similarity_value</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;similarity&#39;</span><span class="p">]</span>
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">similarity_value</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">similarity_value</span> <span class="o">&lt;</span> <span class="mi">98</span><span class="p">:</span>
                    <span class="n">row_cells</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">runs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">font</span><span class="o">.</span><span class="n">highlight_color</span> <span class="o">=</span> <span class="n">WD_COLOR_INDEX</span><span class="o">.</span><span class="n">YELLOW</span>  <span class="c1"># Highlight if less than 98%</span>

            <span class="c1"># Mean coverage depth column</span>
            <span class="n">mean_coverage_depth_value</span> <span class="o">=</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Mean coverage depth&#39;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;Mean coverage depth&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;N/A&quot;</span><span class="p">:</span>
                <span class="n">row_cells</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">text</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mean_coverage_depth_value</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="c1"># Increment figure and table count</span>
        <span class="n">table_count</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">role</span> <span class="o">!=</span> <span class="s2">&quot;superuser&quot;</span><span class="p">:</span>
            <span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;project/static/genomes/EPI_ISL_412866/EPI_ISL_412866.fasta&quot;</span> <span class="k">if</span> <span class="n">sample_reference_map</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;EPI_ISL_412866&quot;</span> <span class="k">else</span> <span class="s2">&quot;project/static/genomes/EPI_ISL_1653999/EPI_ISL_1653999.fasta&quot;</span>
            <span class="n">region</span><span class="o">=</span><span class="s2">&quot;EPI_ISL_412866:1-15225&quot;</span> <span class="k">if</span> <span class="n">sample_reference_map</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;EPI_ISL_412866&quot;</span> <span class="k">else</span> <span class="s2">&quot;EPI_ISL_1653999:1-15222&quot;</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="s2">&quot;project/static/genomes/EPI_ISL_412866/EPI_ISL_412866.gtf&quot;</span> <span class="k">if</span> <span class="n">sample_reference_map</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;EPI_ISL_412866&quot;</span> <span class="k">else</span> <span class="s2">&quot;project/static/genomes/EPI_ISL_1653999/EPI_ISL_1653999.gtf&quot;</span>
            <span class="n">output_dir</span> <span class="o">=</span> <span class="s2">&quot;project/static/plots&quot;</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">reference_genome</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>
            <span class="n">bigwig_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;data/</span><span class="si">{</span><span class="n">distribution</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.bw&quot;</span><span class="p">)</span>
            <span class="n">bigwig_copy</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.bw&quot;</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">bigwig_file_path</span><span class="p">,</span><span class="n">bigwig_copy</span><span class="p">)</span>
            <span class="n">bigwig_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;data/</span><span class="si">{</span><span class="n">distribution</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">_consensus.bw&quot;</span><span class="p">)</span>
            <span class="n">bigwig_consensus_copy</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">_consensus.bw&quot;</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">bigwig_file_path</span><span class="p">,</span><span class="n">bigwig_consensus_copy</span><span class="p">)</span>
            <span class="n">bed_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;data/</span><span class="si">{</span><span class="n">distribution</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">_mutations.bed&quot;</span><span class="p">)</span>
            <span class="n">bed_path_copy</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">user_lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">_mutations.bed&quot;</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">bed_path</span><span class="p">,</span><span class="n">bed_path_copy</span><span class="p">)</span>
            <span class="n">output_dir</span> <span class="o">=</span> <span class="s2">&quot;project/static/plots&quot;</span>
            <span class="n">plot_path</span> <span class="o">=</span> <span class="n">create_pygenometracks_plot</span><span class="p">(</span> <span class="n">reference_genome</span><span class="p">,</span> <span class="n">annotation</span><span class="p">,</span> <span class="n">region</span><span class="p">,</span> <span class="n">bed_path_copy</span><span class="p">,</span> <span class="n">bigwig_copy</span><span class="p">,</span> <span class="n">bigwig_consensus_copy</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">user_lab</span><span class="p">)</span>
            
            <span class="c1"># Add the plot image to DOCX with a caption</span>
            <span class="n">doc</span><span class="o">.</span><span class="n">add_picture</span><span class="p">(</span><span class="n">plot_path</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">Inches</span><span class="p">(</span><span class="mf">7.5</span><span class="p">))</span>  <span class="c1"># Adjust size as needed</span>
            <span class="n">last_paragraph</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> 
            <span class="n">last_paragraph</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
            <span class="n">last_paragraph</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">para</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
            <span class="n">runner</span> <span class="o">=</span> <span class="n">para</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Figure </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">figure_count</span><span class="p">)</span><span class="si">}</span><span class="s2">. Genomic visualisation of submitted sequence and reads.&quot;</span><span class="p">)</span>
            <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">last_paragraph</span> <span class="o">=</span> <span class="n">doc</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> 
            <span class="n">last_paragraph</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
            <span class="n">last_paragraph</span><span class="o">.</span><span class="n">add_run</span><span class="p">()</span><span class="o">.</span><span class="n">add_break</span><span class="p">(</span><span class="n">WD_BREAK</span><span class="o">.</span><span class="n">PAGE</span><span class="p">)</span>
            <span class="n">figure_count</span><span class="o">+=</span><span class="mi">1</span>
        
        

       
























    <span class="c1"># Appendixes</span>
    <span class="n">appendix1</span><span class="o">=</span><span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Appendix 1: Additional Information</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Samples provided and testing required</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;Original samples distributed by UK NEQAS Microbiology were freeze dried.</span>
<span class="s2"> ➢  Sequencing-only laboratories received samples directly and were instructed to reconstitute with 0.5mL of</span>
<span class="s2">molecular grade water prior to testing.</span>
<span class="s2"> ➢  Samples processed by a primary testing laboratory were forwarded on as RNA / lysate and handled and</span>
<span class="s2">stored according to local procedures / policies.</span>
<span class="s2">Sequencing was carried out according to the laboratory&#39;s normal procedure.</span><span class="se">\n\n</span><span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data submission and analysis</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;Data collection, quality control (QC), storage and analysis to  WHO defined standards and requirements was carried out by University of Cranfield in collaborations with UK NEQAS for Microbiology and Micropathology ltd.</span><span class="se">\n</span><span class="s2">Participants were required to submit FASTA, FASTQ or BAM files which are used to assess quality metrics. FASTQ files are processed preferentially over BAM files. If a FASTQ file is not provided but a BAM file is, it will realigned to reference sequence. Should a FASTA not be provided, a consensus sequence will be automatically generated from the FASTQ/BAM, otherwise the provided FASTA will be assumed to be the consensus sequence.</span><span class="se">\n\n</span><span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Validated results</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;The EQA cases were validated by a national RSV sequencing reference laboratory using Illumina sequencing with 99.9% of the genome covered at 20x. The quality of the reference laboratory sequences was classed as very high. Participants have only been assessed against regions successfully covered by the reference laboratory. Stated lineages were established using Nextclade and Nextstrain identifiers are reported. Participant submissions were compared against the reference sequences suggested by GISAID: EPI_ISL_412866 (RSV A) and EPI_ISL_1653999 (RSV B) for all regions successfully sequenced by the reference laboratory.</span><span class="se">\n\n</span><span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Quality Metrics</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span> <span class="o">=</span> <span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;Where appropriate data files have been provided by the participant, the following quality metrics will be stated on the reports:</span>
<span class="s2">➢  Genome Coverage (%) - The percentage of reference bases covered, with a threshold typically set at ≥90%. Computed using Nextclade.</span>
<span class="s2">➢  Ns in Sequence (%) - The percentage of ambiguous bases (N&#39;s) in the sequence, with a threshold typically set at ≤2%. Computed using Nextclade.</span>
<span class="s2">➢  Similarity (%) - The percentage of sequence similarity compared to the reference, with a threshold typically set at ≥98%. Computed using Nextclade.</span>
<span class="s2">➢  Read Coverage (Mean Depth) - The average depth of sequencing reads, with a threshold typically set at ≥50. Computed using Qualimap2.</span><span class="se">\n\n</span><span class="s2">&quot;&quot;&quot;</span><span class="p">)</span> 
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Participation and scoring submissions</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;xxxxxxxxxxxxxxxxxx</span><span class="se">\n\n</span><span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Scheme compliance</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">runner</span><span class="o">=</span><span class="n">appendix1</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;All participating laboratories complied with Scheme instructions.</span><span class="se">\n\n</span><span class="s2">&quot;&quot;&quot;</span><span class="p">)</span>

    <span class="c1">#add generator mention</span>
    <span class="n">current_year</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">year</span>
    <span class="n">automention</span><span class="o">=</span><span class="n">doc</span><span class="o">.</span><span class="n">add_paragraph</span><span class="p">()</span> 
    <span class="n">automention</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
    <span class="n">rights</span><span class="o">=</span><span class="n">automention</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n\n\n\n\n\n\n\n\n\n</span><span class="s2">Automatically generated by xxx. All rights reserved © UKNEQAS for Microbiology; </span><span class="si">{</span><span class="n">current_year</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">rights</span><span class="o">.</span><span class="n">bold</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">rights</span><span class="o">.</span><span class="n">italic</span> <span class="o">=</span> <span class="kc">True</span>
    


    <span class="c1"># Save the document to a BytesIO object and add footers</span>
    <span class="n">docx_io</span> <span class="o">=</span> <span class="n">BytesIO</span><span class="p">()</span>
    <span class="n">add_page_number</span><span class="p">(</span><span class="n">doc</span><span class="o">.</span><span class="n">sections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">footer</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">add_run</span><span class="p">(</span><span class="s2">&quot;RSV Sequencing Distributions ... Pilot EQA Summary Report</span><span class="se">\t\t</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="n">doc</span><span class="o">.</span><span class="n">sections</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">footer</span><span class="o">.</span><span class="n">paragraphs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">WD_PARAGRAPH_ALIGNMENT</span><span class="o">.</span><span class="n">CENTER</span>
    <span class="n">doc</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">docx_io</span><span class="p">)</span>
    <span class="n">docx_io</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">doc</span></div>

</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">rsv-neqas</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../manage.html">Backend management</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">Backend endpoints</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html#backend-utils">Backend utils</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Kevin.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.2</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
    </div>

    

    
  </body>
</html>