<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>project.utils.report_parser &#8212; rsv-neqas 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css?v=27fed22d" />
    <script src="../../../_static/documentation_options.js?v=d45e8c67"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for project.utils.report_parser</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">csv</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">csv</span>

<div class="viewcode-block" id="parse_nextclade_file">
<a class="viewcode-back" href="../../../modules.html#project.utils.report_parser.parse_nextclade_file">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_nextclade_file</span><span class="p">(</span><span class="n">nextclade_file_path</span><span class="p">,</span> <span class="n">nextclade_alternative_file_path</span><span class="p">,</span> <span class="n">genomeLength</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parses Nextclade output and calculates uniformity based on coverage and mutations.&quot;&quot;&quot;</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">nextclade_file_path</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Nextclade file missing: </span><span class="si">{</span><span class="n">nextclade_file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s1">&#39;seqName&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;coverage&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;totalMissing&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;Ns&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s1">&#39;substitutions&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;deletions&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;insertions&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s1">&#39;frameShifts&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;similarity&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;clade&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;G_clade&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;subtype&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span>
        <span class="p">}</span>
    
    <span class="c1"># Initialize the score from alternative if provided, default to None if not provided</span>
    <span class="n">score_alternative</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">nextclade_alternative_file_path</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">nextclade_alternative_file_path</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">nextclade_alternative_file_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
                    <span class="n">score_alternative</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;qc.overallScore&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>  <span class="c1"># Use 0 as default if key is missing</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing alternative file </span><span class="si">{</span><span class="n">nextclade_alternative_file_path</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">score_alternative</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">subtype</span> <span class="o">=</span> <span class="s1">&#39;unknown&#39;</span>  <span class="c1"># Default subtype is unknown</span>
    
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">nextclade_file_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">score</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;qc.overallScore&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>  <span class="c1"># Default to 0 if missing or invalid</span>

                <span class="c1"># Logic to compare scores</span>
                <span class="k">if</span> <span class="n">score_alternative</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">score</span> <span class="o">&lt;</span> <span class="n">score_alternative</span><span class="p">:</span>
                        <span class="n">subtype</span> <span class="o">=</span> <span class="s2">&quot;original&quot;</span>
                    <span class="k">elif</span> <span class="n">score</span> <span class="o">&gt;</span> <span class="n">score_alternative</span><span class="p">:</span>
                        <span class="n">subtype</span> <span class="o">=</span> <span class="s2">&quot;alternative&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">subtype</span> <span class="o">=</span> <span class="s2">&quot;unknown&quot;</span>  <span class="c1"># If scores are equal, mark as unknown</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># If no alternative file, default to original</span>
                    <span class="n">subtype</span> <span class="o">=</span> <span class="s2">&quot;original&quot;</span> <span class="k">if</span> <span class="n">score</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;unknown&quot;</span>

                <span class="c1"># Calculate other necessary values</span>
                <span class="n">coverage</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;coverage&#39;</span><span class="p">])</span>
                <span class="n">total_missing</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;totalMissing&#39;</span><span class="p">])</span>
                <span class="n">alignment_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;alignmentStart&#39;</span><span class="p">])</span>
                <span class="n">alignment_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;alignmentEnd&#39;</span><span class="p">])</span>
                <span class="n">alignment_length</span> <span class="o">=</span> <span class="n">alignment_end</span> <span class="o">-</span> <span class="n">alignment_start</span>
                
                <span class="n">substitutions</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;totalSubstitutions&#39;</span><span class="p">])</span>
                <span class="n">deletions</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;totalDeletions&#39;</span><span class="p">])</span>
                <span class="n">insertions</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;totalInsertions&#39;</span><span class="p">])</span>
                <span class="n">frame_shifts</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;totalFrameShifts&#39;</span><span class="p">])</span>

                <span class="n">clade</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;clade&#39;</span><span class="p">,</span> <span class="s1">&#39;N/A&#39;</span><span class="p">)</span>
                <span class="n">g_clade</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;G_clade&#39;</span><span class="p">,</span> <span class="s1">&#39;N/A&#39;</span><span class="p">)</span>

                <span class="n">adjusted_coverage</span> <span class="o">=</span> <span class="n">alignment_length</span> <span class="o">-</span> <span class="n">total_missing</span> <span class="o">-</span> <span class="n">substitutions</span> <span class="o">-</span> <span class="n">deletions</span> <span class="o">-</span> <span class="n">insertions</span> <span class="o">-</span> <span class="n">frame_shifts</span>
                <span class="n">similarity</span> <span class="o">=</span> <span class="n">adjusted_coverage</span> <span class="o">/</span> <span class="n">genomeLength</span>

                <span class="k">return</span> <span class="p">{</span>
                    <span class="s1">&#39;seqName&#39;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;seqName&#39;</span><span class="p">],</span>
                    <span class="s1">&#39;coverage&#39;</span><span class="p">:</span> <span class="n">coverage</span><span class="p">,</span>
                    <span class="s1">&#39;totalMissing&#39;</span><span class="p">:</span> <span class="n">total_missing</span><span class="p">,</span>
                    <span class="s1">&#39;Ns&#39;</span><span class="p">:</span> <span class="n">total_missing</span> <span class="o">/</span> <span class="n">genomeLength</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
                    <span class="s1">&#39;substitutions&#39;</span><span class="p">:</span> <span class="n">substitutions</span><span class="p">,</span>
                    <span class="s1">&#39;deletions&#39;</span><span class="p">:</span> <span class="n">deletions</span><span class="p">,</span>
                    <span class="s1">&#39;insertions&#39;</span><span class="p">:</span> <span class="n">insertions</span><span class="p">,</span>
                    <span class="s1">&#39;frameShifts&#39;</span><span class="p">:</span> <span class="n">frame_shifts</span><span class="p">,</span>
                    <span class="s1">&#39;similarity&#39;</span><span class="p">:</span> <span class="n">similarity</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
                    <span class="s1">&#39;clade&#39;</span><span class="p">:</span> <span class="n">clade</span><span class="p">,</span>
                    <span class="s1">&#39;G_clade&#39;</span><span class="p">:</span> <span class="n">g_clade</span><span class="p">,</span>
                    <span class="s1">&#39;subtype&#39;</span><span class="p">:</span> <span class="n">subtype</span>
                <span class="p">}</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing Nextclade file </span><span class="si">{</span><span class="n">nextclade_file_path</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">{</span>
            <span class="s1">&#39;seqName&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;coverage&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;totalMissing&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;Ns&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s1">&#39;substitutions&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;deletions&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;insertions&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s1">&#39;frameShifts&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;similarity&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;clade&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;G_clade&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="s1">&#39;subtype&#39;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span>
        <span class="p">}</span></div>


<div class="viewcode-block" id="parse_qualimap">
<a class="viewcode-back" href="../../../modules.html#project.utils.report_parser.parse_qualimap">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_qualimap</span><span class="p">(</span><span class="n">genome_results_path</span><span class="p">,</span> <span class="n">coverage_histogram_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; Parses Qualimap genome_results.txt and coverage histogram file. &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">genome_results_path</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">coverage_histogram_path</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Missing Qualimap files (</span><span class="si">{</span><span class="n">genome_results_path</span><span class="si">}</span><span class="s2"> or </span><span class="si">{</span><span class="n">coverage_histogram_path</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;Coverage at 20X (%)&quot;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s2">&quot;Standard deviation of coverage depth&quot;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s2">&quot;Read depth (Median)&quot;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span><span class="p">,</span>
            <span class="s2">&quot;Uniformity (%)&quot;</span><span class="p">:</span> <span class="s1">&#39;N/A&#39;</span>
        <span class="p">}</span>

    <span class="n">metrics</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome_results_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;There is a&quot;</span> <span class="ow">in</span> <span class="n">line</span> <span class="ow">and</span> <span class="s2">&quot;of reference with a coverageData &gt;= 20X&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Coverage at 20X (%)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;%&#39;</span><span class="p">))</span>

            <span class="k">if</span> <span class="s2">&quot;mean coverageData&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Mean coverage depth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>

            <span class="k">if</span> <span class="s2">&quot;std coverageData&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Standard deviation of coverage depth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">histogram_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">coverage_histogram_path</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">depths</span> <span class="o">=</span> <span class="n">histogram_data</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">histogram_data</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">total_locations</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>

        <span class="n">cumulative_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
        <span class="n">median_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">cumulative_counts</span><span class="p">,</span> <span class="n">total_locations</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Read depth (Median)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="n">median_index</span><span class="p">]</span>

        <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Read depth (Median)&quot;</span><span class="p">]</span>
        <span class="n">uniform_bases</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">depths</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">])</span>
        <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Uniformity (%)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">uniform_bases</span> <span class="o">/</span> <span class="n">total_locations</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing Qualimap histogram: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Read depth (Median)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;N/A&#39;</span>
        <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;Uniformity (%)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;N/A&#39;</span>

    <span class="k">return</span> <span class="n">metrics</span></div>


<div class="viewcode-block" id="read_genome_length">
<a class="viewcode-back" href="../../../modules.html#project.utils.report_parser.read_genome_length">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">read_genome_length</span><span class="p">(</span><span class="n">sample_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; Reads genome length from genomeLength.txt, handling missing file cases. &quot;&quot;&quot;</span>
    <span class="n">genome_length_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="s2">&quot;genomeLength.txt&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">genome_length_file</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome_length_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Invalid genomeLength.txt format in </span><span class="si">{</span><span class="n">sample_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: genomeLength.txt not found in </span><span class="si">{</span><span class="n">sample_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="process_all_reports">
<a class="viewcode-back" href="../../../modules.html#project.utils.report_parser.process_all_reports">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">process_all_reports</span><span class="p">(</span><span class="n">base_dir</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; Processes all samples in the given directory and extracts relevant metrics. &quot;&quot;&quot;</span>

    <span class="n">report_data</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">base_dir</span><span class="p">):</span>
        <span class="n">lab_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="n">lab</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">lab_path</span><span class="p">):</span>
            <span class="n">report_data</span><span class="p">[</span><span class="n">lab</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">lab_path</span><span class="p">):</span>
                <span class="n">sample_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">lab_path</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">sample_path</span><span class="p">):</span>
                    <span class="n">genome_results_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="s2">&quot;genome_results.txt&quot;</span><span class="p">)</span>
                    <span class="n">coverage_histogram_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="s2">&quot;raw_data_qualimapReport/coverage_histogram.txt&quot;</span><span class="p">)</span>
                    <span class="n">nextclade_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="s2">&quot;nextclade.output&quot;</span><span class="p">)</span>
                    <span class="n">nextclade_alternative_file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="s2">&quot;nextclade_alternative.output&quot;</span><span class="p">)</span>

                    <span class="c1"># Read genome length</span>
                    <span class="n">genome_length</span> <span class="o">=</span> <span class="n">read_genome_length</span><span class="p">(</span><span class="n">sample_path</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">genome_length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="k">continue</span>  <span class="c1"># Skip if genome length is missing</span>

                    <span class="n">report_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                        <span class="s1">&#39;fasta&#39;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.fasta&quot;</span><span class="p">),</span>
                        <span class="s1">&#39;bam&#39;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.bam&quot;</span><span class="p">),</span>
                        <span class="s1">&#39;bai&#39;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_path</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">lab</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.bam.bai&quot;</span><span class="p">),</span>
                    <span class="p">}</span>

                    <span class="c1"># Process Qualimap data</span>
                    <span class="n">qualimap_metrics</span> <span class="o">=</span> <span class="n">parse_qualimap</span><span class="p">(</span><span class="n">genome_results_path</span><span class="p">,</span> <span class="n">coverage_histogram_path</span><span class="p">)</span>
                    <span class="n">report_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="n">sample</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">qualimap_metrics</span><span class="p">)</span>

                    <span class="c1"># Process Nextclade data</span>
                    <span class="n">nextclade_metrics</span> <span class="o">=</span> <span class="n">parse_nextclade_file</span><span class="p">(</span><span class="n">nextclade_file_path</span><span class="p">,</span> <span class="n">nextclade_alternative_file_path</span><span class="p">,</span> <span class="n">genome_length</span><span class="p">)</span>
                    <span class="n">report_data</span><span class="p">[</span><span class="n">lab</span><span class="p">][</span><span class="n">sample</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">nextclade_metrics</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">report_data</span></div>

</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">rsv-neqas</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">project</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html#project-utils">project.utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html#frontend-documentation">Frontend Documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  <li><a href="../../project.html">project</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Kevin.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.2</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
    </div>

    

    
  </body>
</html>